Moving WormManager from d2k to Common as ActorSpawnManager

CDVoidwalker opened the pull request:
Works just like old WormManager with addition of being conditional and having Types.

MustaphaTR commented:
> Checking for explicit interface violations...
> ActorSpawnManager must explicitly implement the interface member ITick.Tick
> Explicit interface violations: 1
> make: *** [check] Error 1

Also commits should be squashed.

GraionDilach commented:
>OpenRA.Mods.Common/Traits/ActorSpawner.cs:L25: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.
OpenRA.Mods.Common/Traits/ActorSpawner.cs:L14: [SA1208] System using directives must be placed before all other using directives.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L21: [SA1001] Invalid spacing around the comma.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L34: [SA1012] Invalid spacing around the opening curly bracket.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L34: [SA1013] Invalid spacing around the closing curly bracket.

Unit158 commented:
There's a binary file that shouldn't have been included.

Unit158 commented on mods/d2k/rules/world.yaml:
```diff
@@ -72,7 +72,9 @@ World:
 		ProductionPaletteWidget: PRODUCTION_PALETTE
-	WormManager:
+	ActorSpawnManager:
+		ActorSignature: sandworm
```
Why not just Actor:

Unit158 commented on mods/d2k/rules/world.yaml:
```diff
@@ -72,7 +72,9 @@ World:
-	WormManager:
+	ActorSpawnManager:
+		ActorSignature: sandworm
+		ActorSignNotification: WormSign
```
ActorSpawnNotification:?

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+		public readonly string[] ActorSignature = { };
+
+		public readonly string ActorSignNotification = "";
+		public readonly string ActorOwnerPlayer = "Creeps";
```
I'd argue that this would be just as clear as "ActorOwner:"

CDVoidwalker replied:
That was directly copied from WormManager, as with all the comments

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+		public readonly string ActorSignNotification = "";
+		public readonly string ActorOwnerPlayer = "Creeps";
+
+		[Desc("Type of ActorSpawner which it connects with")]
```
"which it connects with" -> "with which it connects"

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+
+		bool enabled;
+		int spawnCountdown;
+		int actorPresent;
```
actorsPresent

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+
+			do
+			{
+				// Always spawn at least one actor, plus however many
```
"Always spawn at least one actor, plus however many needed to reach the minimum"

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+			if (!spawnPointActors.Value.Any())
+				return;
+
+			// Apparently someone doesn't want actors or the maximum number of actors has been reached
```
I'd argue the next line doesn't warrant a comment.

Unit158 commented on OpenRA.Mods.Common/Traits/ActorSpawner.cs:
```diff
@@ -0,0 +1,34 @@
+	[Desc("An actor with this trait indicates a valid spawn point for actors.")]
+	public class ActorSpawnerInfo : ConditionalTraitInfo
+	{
+		[Desc("Type of ActorSpawner which it connects with")]
```
"which it connects with" -> "with which it connects"

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] ActorSignature = { };
```
ActorSignature -> Actors

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] ActorSignature = { };
+
+		public readonly string ActorSignNotification = "";
```
SpawnNotification

penev92 replied:
No, that was actually `public readonly string WormSignNotification = "WormSign";`.
So this one probably needs some more work.

CDVoidwalker replied:
[FieldLoader.Require] ? or null?

Unit158 commented:
I haven't tested yet, but the code looks OK, minus some style/naming/language improvements.

Unit158 commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,119 @@
+			return spawnPoint.CenterPosition;
+		}
+
+		Actor GetRandomSpawnPoint(Actor self)
```
I'd argue this should only pass the shared RNG.

MunWolf commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+			if (info.Maximum < 1 || actorsPresent >= info.Maximum)
+				return;
+
+			if (--spawnCountdown > 0 && actorsPresent >= info.Minimum)
```
This might not ever happen but there could be an "underflow" here if the second condition never fires.

CDVoidwalker replied:
There's no real case when it would happen, as adding actors is after this check

MunWolf replied:
I meant that --spawnCountdown might eventually underflow to a really big value after -2147483648 ticks... which will probably never happen since it will take way too long.

reaperrr replied:
@MunWolf Unless I'm miscalculating, we're talking about nearly 12000 hours at maximum game speed (50 ticks/sec).

I dare say whoever manages to set things up weirdly enough to trigger this, deserves the crash ;)

MunWolf commented:
This still needs an upgrade rule for the name change.

pchote commented:
Can you please also squash the fixup commits into the main one?

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -837,22 +839,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/common/"/>
```
Please unstage these unrelated project file formatting changes.

pchote commented on OpenRA.Mods.Common/Traits/ActorSpawner.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("An actor with this trait indicates a valid spawn point for actors.")]
```
Can you please have this mention that it is specifically for `ActorSpawnManager`?

pchote commented on OpenRA.Mods.Common/Traits/ActorSpawner.cs:
```diff
@@ -0,0 +1,34 @@
+	public class ActorSpawner : ConditionalTrait<ActorSpawnerInfo>
+	{
+		public ActorSpawner(ActorSpawnerInfo info) : base(info)
+		{ }
```
Our code style conventions for cases like this is
```csharp
public ActorSpawner(ActorSpawnerInfo info)
	: base(info) { }
```

pchote commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls the spawning of actors. Attach this to the world actor.")]
```
Can you please change this to something like "Controls spawning of map creep actors"?  There are other types of actor spawning in the game, so this could otherwise be confusing.

CDVoidwalker replied:
Yeah, but also those actors aren't creep only, depends on what is set in Owner of the manager, so "Controls spawning of the map actors"?

pchote replied:
I would go with "Controls the spawning of specified actor types" to keep the distinction vs e.g. `SpawnMpUnits` clear.

pchote commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] Actors = { };
+
+		public readonly string ActorSignNotification = "";
```
This isn't used, so please remove it.

pchote commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] Actors = { };
```
These are actor references, so please attach the `[ActorReference]` annotation.

pchote commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
```
Please mention that the spawned actor types are selected randomly from this list - otherwise it could be interpreted as spawning them all at once.

pchote commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,118 @@
+
+			spawnCountdown = info.SpawnInterval;
+
+			var actorLocations = new List<WPos>();
```
This isn't used for anything?

pchote commented on OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj:
```diff
@@ -121,22 +119,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/d2k/"/>
```
Please unstage unrelated project file changes.

pchote reviewed (commented):
I haven't tested functionality, but I have left a few style nits.
As @MunWolf mentioned this will also need an upgrade rule to manage the rename.

pchote commented:
It would be good to keep this moving and to get it merged before it loses momentum.

Do you need any help with this?

CDVoidwalker commented:
Will try to get it done for tomorrow

GraionDilach commented:
>OpenRA.Mods.Common/Traits/ActorSpawner.cs:L29: [SP2000] Invalid spacing at the end of the line.

CDVoidwalker commented:
needs review

MustaphaTR commented:
Needs rebase.

GraionDilach reviewed (approved):
:+1: but someone will need to update the date of the upgrade rule.

abcdefg30 commented:
Also needs a rebase yet again. :/

CDVoidwalker committed:
Generalize WormManager into ActorSpawnManager.

Added support of multiple actors, conditions and types.

pchote commented:
I want to see this merged for the playtest, so I have amended the branch with a few changes:
* Rebased and updated upgrade rule date.
* Replaced the default "" type with the same "overlaps or types is empty" check that we use everywhere else.
* Replaced the `Lazy<TraitPair<ActorSpawner>[]>` with a more modern Created lookup.

pchote reviewed (dismissed):
LGTM and works as advertised ingame.

My amendments will need a :+1: before merging.

GraionDilach commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,116 @@
+	[Desc("Controls the spawning of specified actor types. Attach this to the world actor.")]
+	public class ActorSpawnManagerInfo : ConditionalTraitInfo, Requires<MapCreepsInfo>
+	{
+		[Desc("Minimum number of actors")]
```
Missing fullstop.

GraionDilach commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,116 @@
+		[Desc("Minimum number of actors")]
+		public readonly int Minimum = 0;
+
+		[Desc("Maximum number of actors")]
```
Missing fullstop.

GraionDilach commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,116 @@
+
+		Actor GetRandomSpawnPoint(Support.MersenneTwister random)
+		{
+			return spawnPointActors.Random(random).Actor;
```
ActorSpawner traits are also conditional, but this is ignored in this current revision. IMO it'd make the most sense to filter the spawnpoints here.

GraionDilach commented:
I'm actually not certain how much sense it makes to cache the suitable actor spawners on creation. Not doing that would allow one to enable map creeps midgame based on some initial condition.

pchote commented:
Yes, that makes sense.  Certain mods or minigames might want to use player-built actors as spawn points.

Arular101 commented:
Please don't forget to update the copyright notice year.

reaperrr committed:
Don't cache ActorSpawners at creation

Re-evaluate before every spawn attempt instead.
Also accounts for ActorSpawner being conditional now.

While a bit more costly in terms of performance, this allows to create spawns mid-game.

reaperrr commented:
Adressed all remaining style nits, replaced caching-at-creation with evaluation right before spawn attempt, rebased.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs(48,29): error CS0169: Warning as Error: The private field `OpenRA.Mods.Common.Traits.ActorSpawnManager.spawnPointActors' is never used

reaperrr commented:
Fixed.

GraionDilach commented on OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:
```diff
@@ -0,0 +1,112 @@
+
+			var spawnPoint = GetRandomSpawnPoint(self.World, self.World.SharedRandom);
+
+			if (!enabled)
```
Does caching the nullcheck to a boolean for this usage worth it?

reaperrr commented:
Updated.

> Does caching the nullcheck to a boolean for this usage worth it?

No, but I also forgot to keep the MapCreeps.Enabled check, so now I'm using `enabled` to cache that in the ctor (ASM Requires MapCreeps, so the latter should be constructed before the former) but use `spawnPoint == null` to abort later if there's no enabled spawnpoint.

reaperrr commented:
I've proof-read and tested my own updates. If it turns out I overlooked anything, I'll take responsibility and file a follow-up.

GraionDilach commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1635,6 +1635,21 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20180225)
```
Needs an update on this due to new release (?).

reaperrr replied:
Yeah, but it's not the only case, I think. I'll fix it either way.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3a17c2590602a423b33f9e6ebbb9e36d62dcb982...ad98895675fe2bee9a7000cd607b74bffe669928
